tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Testing/MsTestV2/TestMethodAttributeExecuteIntegration.cs

```diff
@@ -56,80 +74,99 @@ public static class TestMethodAttributeExecuteIntegration
+                                test.SetErrorInfo(testExceptionType.FullName ?? "Error", testException.Message, testException.ToString());
                             }
+                        }
 
-                            break;
-                        case UnitTestOutcome.Passed:
-                            if (exception is not null)
-                            {
-                                test.SetErrorInfo(exception);
-                                test.Close(TestStatus.Fail);
-                            }
-                            else
+                        if (!string.IsNullOrEmpty(testResult.DisplayName) && test.Name != testResult.DisplayName)
+                        {
+                            test.SetName(testResult.DisplayName);
+                            MsTestIntegration.UpdateTestParameters(test, testMethodState.TestMethod, testResult.DisplayName);
+                        }
+
+                        if (exception is not null)
+                        {
+                            test.Close(TestStatus.Fail);
+                        }
+                        else
+                        {
+                            switch (testResult.Outcome)
                             {
-                                test.Close(TestStatus.Pass);
+                                case UnitTestOutcome.Error or UnitTestOutcome.Failed or UnitTestOutcome.Timeout:
+                                    test.Close(TestStatus.Fail);
+                                    break;
+                                case UnitTestOutcome.Inconclusive or UnitTestOutcome.NotRunnable:
+                                    test.Close(TestStatus.Skip, TimeSpan.Zero, testException?.Message ?? string.Empty);
+                                    break;
+                                case UnitTestOutcome.Passed:
+                                    test.Close(TestStatus.Pass);
+                                    break;
+                                default:
+                                    Common.Log.Warning("Failed to handle the test status");
+                                    test.Close(TestStatus.Fail);
+                                    break;
                             }
-
-                            break;
+                        }
+                    }
+                    else
+                    {
+                        Common.Log.Warning("Failed to cast TestResultStruct");
+                        test.Close(TestStatus.Fail);
                     }
                 }
             }
+            else
+            {
+                Common.Log.Warning("Failed to extract TestResult from return value");
+                testMethodState.Test.Close(TestStatus.Fail);
+            }
         }
 
         return new CallTargetReturn<TReturn>(returnValue);
     }
+
+    private class TestMethodState
+    {
+        public TestMethodState(Test test, ITestMethod testMethod)
```

Comment:
This will box `ITestMethod` right, can/should we make `TestMethodState` generic?

Reply:
Sadly we can't ,`TTestMethod` is a parameter constraint in `OnMethodBegin` we don't have the same constraint in `OnMethodEnd`